src/middlewared/middlewared/plugins/zettarepl.py

```diff
@@ -511,22 +572,69 @@ def _observer_queue_reader(self):
                             "total": message.total,
                         }
                     }
+
+                    channel = self.replication_jobs_channels.get(message.task_id)
+                    if channel:
+                        channel.put(message)
+                    else:
+                        self.logger.warning("Don't have job channel for task %r", message.task_id)
+
                 if isinstance(message, ReplicationTaskSnapshotSuccess):
                     self.last_snapshot[f"replication_{message.task_id}"] = f"{message.dataset}@{message.snapshot}"
+
+                    channel = self.replication_jobs_channels.get(message.task_id)
+                    if channel:
+                        channel.put(message)
+                    else:
+                        self.logger.warning("Don't have job channel for task %r", message.task_id)
+
                 if isinstance(message, ReplicationTaskSuccess):
                     self.state[f"replication_{message.task_id}"] = {
                         "state": "FINISHED",
                         "datetime": datetime.utcnow(),
                     }
+
+                    channel = self.replication_jobs_channels.pop(message.task_id, None)
+                    if channel:
+                        channel.put(message)
+                    else:
+                        self.logger.warning("Don't have job channel for task %r", message.task_id)
+
                 if isinstance(message, ReplicationTaskError):
                     self.state[f"replication_{message.task_id}"] = {
                         "state": "ERROR",
                         "datetime": datetime.utcnow(),
                         "error": message.error,
                     }
+
+                    channel = self.replication_jobs_channels.pop(message.task_id, None)
+                    if channel:
+                        channel.put(message)
+                    else:
+                        self.logger.warning("Don't have job channel for task %r", message.task_id)
+
             except Exception:
                 self.logger.warning("Unhandled exception in observer_queue_reader", exc_info=True)
 
+    @job(logs=True)
+    def replication_job(self, job, task_id):
```

Comment:
Thats awesome!

One thing though, zettarepl is a private service.
What are your thoughts of reusing `replication.run` to do it, turning it into a job?

That way we could call it just like any other method and if there isnt a repl job channel we start one. Possible?